mods/carts/init.lua

```diff
@@ -0,0 +1,413 @@
+		self.object:setvelocity(vel)
+		self.old_dir.y = 0
+	elseif vector.equals(vel, {x=0, y=0, z=0}) then
+		rail_sound(self)
+		rail_on_step_event(railparams.on_step, self, dtime)
+		return
+	end
+
+	-- stop cart if velocity vector flips
+	if self.old_vel and (((self.old_vel.x * vel.x) < 0) or
+			((self.old_vel.z * vel.z) < 0)) then
+		self.old_dir = vector.new()
+		self.old_pos = pos
+		self.object:setvelocity(vector.new())
+		self.object:setacceleration(vector.new())
+		rail_sound(self)
+		rail_on_step_event(railparams.on_step, self, dtime)
+		return
+	end
+	self.old_vel = vector.new(vel)
+
+	if self.old_pos and not self.punched then
+		local flo_pos = vector.round(pos)
+		local flo_old = vector.round(self.old_pos)
+		if vector.equals(flo_pos, flo_old) then
+			rail_sound(self)
+			-- Do not check one node multiple times
+			rail_on_step_event(railparams.on_step, self, dtime)
+			return
+		end
+	end
+
+	local ctrl, player
+
+	-- Get player controls
+	if self.driver then
+		player = minetest.get_player_by_name(self.driver)
+		if player then
+			ctrl = player:get_player_control()
+		end
+	end
+
+	if self.old_pos then
+		-- Detection for "skipping" nodes
+		local expected_pos = vector.add(self.old_pos, self.old_dir)
+		local found_path = carts:pathfinder(
+			pos, expected_pos, self.old_dir, ctrl, self.old_switch, self.railtype
+		)
+
+		if not found_path then
+			-- No rail found: reset back to the expected position
+			pos = vector.new(self.old_pos)
+			update.pos = true
+		end
+	end
+
+	local cart_dir = carts:velocity_to_dir(vel)
+
+	-- dir:         New moving direction of the cart
+	-- switch_keys: Currently pressed L/R key, used to ignore the key on the next rail node
+	local dir, switch_keys = carts:get_rail_direction(
+		pos, cart_dir, ctrl, self.old_switch, self.railtype
+	)
+
+	local new_acc = {x=0, y=0, z=0}
+	if vector.equals(dir, {x=0, y=0, z=0}) then
+		vel = {x=0, y=0, z=0}
+		pos = vector.round(pos)
+		update.pos = true
+		update.vel = true
+	else
+		-- If the direction changed
+		if dir.x ~= 0 and self.old_dir.z ~= 0 then
+			vel.x = dir.x * math.abs(vel.z)
+			vel.z = 0
+			pos.z = math.floor(pos.z + 0.5)
+			update.pos = true
+		end
+		if dir.z ~= 0 and self.old_dir.x ~= 0 then
+			vel.z = dir.z * math.abs(vel.x)
+			vel.x = 0
+			pos.x = math.floor(pos.x + 0.5)
+			update.pos = true
+		end
+		-- Up, down?
+		if dir.y ~= self.old_dir.y then
+			vel.y = dir.y * math.abs(vel.x + vel.z)
+			pos = vector.round(pos)
+			update.pos = true
+		end
+
+		-- Slow down or speed up..
+		local acc = dir.y * -1.8
+
+		-- no need to check for railparams == nil since we always make it exist.
+		local speed_mod = railparams.acceleration
+		if speed_mod and speed_mod ~= 0 then
+			-- Try to make it similar to the original carts mod
+			acc = acc + (speed_mod * 10)
+		else
+			-- Handbrake
+			if ctrl and ctrl.down then
+				acc = acc - 1.6
+			else
+				acc = acc - 0.4
+			end
+		end
+
+		if self.old_dir.y ~= 1 and not self.punched then
```

Comment:
The code within this `if` became useless due addition of https://github.com/minetest/minetest_game/pull/1318/files#diff-9812a7ca8951c9d004454c5c7c339f40R198